File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/safety/date_calendar.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from datetime import date
from typing import Literal

ISO_DATE = re.compile(r"^(\d{4})-(\d{2})-(\d{2})$")

PediatricBand = Literal["P13_A", "P13_B", "P13_C", "P13_D", "P13_E"] | None


@dataclass(frozen=True)
class CalendarDate:
    year: int
    month: int
    day: int


def parse_iso_date_only(value: str | None) -> CalendarDate | None:
    if value is None or not isinstance(value, str):
        return None
    trimmed = value.strip()
    m = ISO_DATE.match(trimmed)
    if not m:
        return None
    year, month, day = int(m.group(1)), int(m.group(2)), int(m.group(3))
    try:
        date(year, month, day)
    except ValueError:
        return None
    return CalendarDate(year, month, day)


def _to_date(parts: CalendarDate) -> date:
    return date(parts.year, parts.month, parts.day)


def calendar_days_between(from_parts: CalendarDate, to_parts: CalendarDate) -> int:
    return (_to_date(to_parts) - _to_date(from_parts)).days


def first_birthday_date(dob: CalendarDate) -> CalendarDate:
    try:
        fb = date(dob.year + 1, dob.month, dob.day)
        return CalendarDate(fb.year, fb.month, fb.day)
    except ValueError:
        fb = date(dob.year + 1, dob.month, max(1, dob.day - 1))
        return CalendarDate(fb.year, fb.month, fb.day)


def is_strictly_before(left: CalendarDate, right: CalendarDate) -> bool:
    return _to_date(left) < _to_date(right)


def completed_years_between(dob: CalendarDate, on: CalendarDate) -> int:
    years = on.year - dob.year
    try:
        birthday_this_year = date(on.year, dob.month, dob.day)
    except ValueError:
        birthday_this_year = date(on.year, dob.month, max(1, dob.day - 1))
    if _to_date(on) < birthday_this_year:
        years -= 1
    return max(0, years)


def band_from_days_and_calendar(
    days_since_birth: int, dob: CalendarDate, assessment: CalendarDate
) -> PediatricBand:
    fb = first_birthday_date(dob)
    if is_strictly_before(assessment, fb):
        if 0 <= days_since_birth <= 28:
            return "P13_A"
        if days_since_birth >= 29:
            return "P13_B"
        return None
    completed = completed_years_between(dob, assessment)
    if 1 <= completed <= 5:
        return "P13_C"
    if 6 <= completed <= 12:
        return "P13_D"
    if completed >= 13:
        return "P13_E"
    return None

File: packages/clinical-contracts/src/rule4/safety/dateCalendar.ts
/** UTC calendar date (no time-of-day) for verified DOB / assessment comparisons. */

export type CalendarDateParts = { year: number; month: number; day: number };

const ISO_DATE = /^(\d{4})-(\d{2})-(\d{2})$/;

export function parseIsoDateOnly(value: string | null | undefined): CalendarDateParts | null {
  if (value == null || typeof value !== 'string') {
    return null;
  }
  const trimmed = value.trim();
  const m = ISO_DATE.exec(trimmed);
  if (!m) {
    return null;
  }
  const year = Number(m[1]);
  const month = Number(m[2]);
  const day = Number(m[3]);
  if (!Number.isInteger(year) || !Number.isInteger(month) || !Number.isInteger(day)) {
    return null;
  }
  if (month < 1 || month > 12 || day < 1 || day > 31) {
    return null;
  }
  const probe = new Date(Date.UTC(year, month - 1, day));
  if (
    probe.getUTCFullYear() !== year ||
    probe.getUTCMonth() !== month - 1 ||
    probe.getUTCDate() !== day
  ) {
    return null;
  }
  return { year, month, day };
}

export function calendarDateToUtcMs(parts: CalendarDateParts): number {
  return Date.UTC(parts.year, parts.month - 1, parts.day);
}

export function calendarDaysBetween(from: CalendarDateParts, to: CalendarDateParts): number {
  const ms = calendarDateToUtcMs(to) - calendarDateToUtcMs(from);
  return Math.floor(ms / 86_400_000);
}

/** First birthday calendar date (same month/day when valid; Feb 29 → Feb 28 on non-leap years). */
export function firstBirthdayDate(dob: CalendarDateParts): CalendarDateParts {
  const targetYear = dob.year + 1;
  const probe = new Date(Date.UTC(targetYear, dob.month - 1, dob.day));
  if (probe.getUTCMonth() === dob.month - 1 && probe.getUTCDate() === dob.day) {
    return { year: targetYear, month: dob.month, day: dob.day };
  }
  return {
    year: targetYear,
    month: dob.month,
    day: Math.max(1, dob.day - 1),
  };
}

export function isStrictlyBeforeCalendar(
  left: CalendarDateParts,
  right: CalendarDateParts,
): boolean {
  return calendarDateToUtcMs(left) < calendarDateToUtcMs(right);
}

export function completedYearsBetween(dob: CalendarDateParts, on: CalendarDateParts): number {
  let years = on.year - dob.year;
  const birthdayThisYear: CalendarDateParts = { year: on.year, month: dob.month, day: dob.day };
  const probe = new Date(Date.UTC(on.year, dob.month - 1, dob.day));
  const effectiveBirthday: CalendarDateParts =
    probe.getUTCMonth() === dob.month - 1 && probe.getUTCDate() === dob.day
      ? birthdayThisYear
      : { year: on.year, month: dob.month, day: Math.max(1, dob.day - 1) };
  if (isStrictlyBeforeCalendar(on, effectiveBirthday)) {
    years -= 1;
  }
  return Math.max(0, years);
}

export function bandFromDaysAndCalendar(
  daysSinceBirth: number,
  dob: CalendarDateParts,
  assessment: CalendarDateParts,
): 'P13_A' | 'P13_B' | 'P13_C' | 'P13_D' | 'P13_E' | null {
  const fb = firstBirthdayDate(dob);
  if (isStrictlyBeforeCalendar(assessment, fb)) {
    if (daysSinceBirth >= 0 && daysSinceBirth <= 28) {
      return 'P13_A';
    }
    if (daysSinceBirth >= 29) {
      return 'P13_B';
    }
    return null;
  }
  const completed = completedYearsBetween(dob, assessment);
  if (completed >= 1 && completed <= 5) {
    return 'P13_C';
  }
  if (completed >= 6 && completed <= 12) {
    return 'P13_D';
  }
  if (completed >= 13) {
    return 'P13_E';
  }
  return null;
}
